File: solutions/clojure/strain/1/src/strain.clj
(ns strain)

(defn retain [p c] ;; <- arglist goes here
  ;; your code goes here
  (loop [c c result []]
    (if (empty? c)
      result
      (let [head (first c) tail (rest c)]
        (recur
         tail
         (if (p head)
           (conj result head)
           result))))))

(defn discard [p c] ;; <- arglist goes here
  ;; your code goes here
  (retain (complement p) c))

File: solutions/go/strain/1/strain.go
package strain

func Keep[T any](input []T, predicate func(t T) bool) (result []T) {
	for _, value := range input {
        if predicate(value) {
            result = append(result, value)
        }
    }
    return result
}

func Discard[T any](input []T, predicate func(t T) bool) []T {
    negate := func(t T) bool {
        return !predicate(t)
    }
    return Keep(input, negate)
}
